### PR TITLE
Show owner of J9VMInternals$ClassInitializationLock in javacore

### PR DESCRIPTION
**Pull request by yathamravali**
Currently javacore shows unowned for the threads waiting on
java/lang/J9VMInternals$ClassInitializationLock. This PR shows
the thread currently responsible for advancing the initialization
of a class using a separate label(Initializing thread).

Fixes: #284
Signed-off-by: Ravali Yatham <ravaliyatham@gmail.com>

### CODE REVIEW

**Comment by yathamravali** [comment]
Performed a test using the testcase https://github.com/eclipse-openj9/openj9/issues/284#issuecomment-704794687

Before:
```
3XMTHREADINFO      "Class Initialization Thread 1" J9VMThread:0x00000000015EEE00, omrthread_t:0x00007FEF6437EB10, java/lang/Thread:0x00000000FFF91138, state:CW, prio=5
3XMJAVALTHREAD            (java/lang/Thread getId:0x15, isDaemon:false)
3XMTHREADINFO1            (native thread ID:0x6DDA, native priority:0x5, native policy:UNKNOWN, vmstate:CW, vm thread flags:0x00000181)
3XMTHREADINFO2            (native stack address range from:0x00007FEF4E13D000, to:0x00007FEF4E17D000, size:0x40000)
3XMCPUTIME               CPU usage total: 0.000117515 secs, current category="Application"
3XMTHREADBLOCK     Waiting on: java/lang/J9VMInternals$ClassInitializationLock@0x00000000FFFC1EF0 Owned by: <unowned>
```

After(with the PR change)
```
3XMTHREADINFO      "Class Initialization Thread 2" J9VMThread:0x000000000015BC00, omrthread_t:0x00007FE828151B88, java/lang/Thread:0x00000000FFF54DF0, state:CW, prio=5
3XMJAVALTHREAD            (java/lang/Thread getId:0x13, isDaemon:false)
3XMJAVALTHRCCL            sun/misc/Launcher$AppClassLoader(0x00000000FFF3C0B0)
3XMTHREADINFO1            (native thread ID:0x6E02, native priority:0x5, native policy:UNKNOWN, vmstate:CW, vm thread flags:0x00000181)
3XMTHREADINFO2            (native stack address range from:0x00007FE8110CE000, to:0x00007FE81110E000, size:0x40000)
3XMCPUTIME               CPU usage total: 0.009084147 secs, current category="Application"
3XMTHREADBLOCK     Waiting on: java/lang/J9VMInternals$ClassInitializationLock@0x00000000FFF5DC90 Owned by: "Class Initialization Thread 1"
```

**Comment by gacholio** [comment]
I don't see anywhere that's distinguishing class init locks from other locks. This code is certainly not valid in general for every lock that's in wait state.

**Comment by gacholio** [comment]
It's also misleading to call this the lock owner in this case. I think this needs to be special cased for class init locks and show a different message.

**Comment by yathamravali** [comment]
@gacholio  Thanks for reviewing.

> It's also misleading to call this the lock owner in this case. I think this needs to be special cased for class init locks and show a different message.

For the above comment, can I add another label(InitializingThread) like below for class init locks?

```
3XMTHREADBLOCK     Waiting on: java/lang/J9VMInternals$ClassInitializationLock@0x00000000FFF5DC90 Owned by: <unowned> InitializingThread: "Class Initialization Thread 1"
```

**Comment by yathamravali** [comment]
@gacholio @keithc-ca Added another commit, PTAL.

**Comment by gacholio** [comment]
I still don't like this being handled as part of lock ownership - the javacore must not contain incorrect or misleading information.

**Comment by gacholio** [comment]
If @keithc-ca thinks this is reasonable, I won't argue.

**Comment by gacholio** [comment]
I was confused by the THREADBLOCK in the header, but I see now that's how it's always been. I'll withdraw my objections.

**Review by gacholio — APPROVED** [review]

File: runtime/rasdump/javadump.cpp [hunk 1]
```diff
@@ -4643,6 +4643,23 @@ JavaCoreDumpWriter::writeThreadBlockers(J9VMThread* vmThread, UDATA vmstate, j9o
 			// Should only occur for WAITING threads.
 			// (For BLOCKED threads this would indicate a broken VM)
 			_OutputStream.writeCharacters("<unowned>");
+
+			// Get J9VMThread details from initializeStatus field.
+			J9Class *lockClass = J9VMJAVALANGJ9VMINTERNALSCLASSINITIALIZATIONLOCK_OR_NULL(vmThread->javaVM);
+			if (NULL != lockClass) {
+				J9Class *clazz = J9OBJECT_CLAZZ(vmThread, lockObject);
```
Comment (keithc-ca) [inline runtime/rasdump/javadump.cpp:4650]: I think `lockObject` may be `NULL` here; please check for that.

File: runtime/rasdump/javadump.cpp [hunk 2]
```diff
@@ -4643,6 +4643,23 @@ JavaCoreDumpWriter::writeThreadBlockers(J9VMThread* vmThread, UDATA vmstate, j9o
...
+				if (instanceOfOrCheckCastNoCacheUpdate(clazz, lockClass)) {
+					j9object_t theClass = J9VMJAVALANGJ9VMINTERNALSCLASSINITIALIZATIONLOCK_THECLASS_VM(vmThread->javaVM, lockObject);
+					J9Class *vmRef = J9VMJAVALANGCLASS_VMREF(vmThread, theClass);
+					J9VMThread *initializingThread = (J9VMThread *)(vmRef->initializeStatus & ~(UDATA)J9ClassInitStatusMask);
+					if (NULL != initializingThread) {
+						_OutputStream.writeCharacters(" InitializingThread: ");
```
Comment (keithc-ca) [inline runtime/rasdump/javadump.cpp:4656]: Please separate the words here:
```
						_OutputStream.writeCharacters(" Initializing thread: ");
```

**Review by keithc-ca — CHANGES_REQUESTED** [review]

File: runtime/rasdump/javadump.cpp [hunk 3]
```diff
@@ -4643,6 +4643,23 @@ JavaCoreDumpWriter::writeThreadBlockers(J9VMThread* vmThread, UDATA vmstate, j9o
...
+						_OutputStream.writeCharacters("\"");
```
Comment (gacholio) [inline runtime/rasdump/javadump.cpp:4657]: This line could be combined with the one above.

File: runtime/rasdump/javadump.cpp (hunk 1 above)
Comment (yathamravali) [reply runtime/rasdump/javadump.cpp]: Thanks, Added condition to check that.

**Comment by yathamravali** [comment]
@keithc-ca - Incorporated suggested changes, PTAL.

**Review by keithc-ca — COMMENTED** [review]
The result looks good.

Please squash and fix your commit message:
* each sentence should end with a period (except the summary line)
* if you use the word "separate", please spell it correctly
* the signed-off-by line should be
    `Signed-off-by: Ravali Yatham <rayatha1@in.ibm.com>`

**Comment by yathamravali** [comment]
Thanks @keithc-ca , squashed the commits.

**Review by keithc-ca — APPROVED** [review]

**Comment by keithc-ca** [comment]
jenkins test sanity amac jdk17

**Comment by pshipton** [comment]
@yathamravali As per https://github.com/eclipse-openj9/openj9/issues/284#issuecomment-1208286785 pls create a doc issue at https://github.com/eclipse-openj9/openj9-docs/issues

**Pull request merged**